File: client/ios/CoinbaseWalletSDK.swift
//
//  CoinbaseWalletSDK.swift
//  WalletSegue
//
//  Created by Jungho Bang on 5/20/22.
//

import Foundation
import CryptoKit

public class CoinbaseWalletSDK {
    let appId: String
    let host: URL
    let callback: URL
    let version: String
    
    lazy var keyManager: KeyManager = {
        KeyManager(host: self.host)
    }()
    lazy var taskManager: TaskManager = {
        TaskManager()
    }()
    
    /// Instantiate Coinbase Wallet SDK
    /// - Parameters:
    ///   - host: universal link url of the host wallet to interact with
    ///   - callback: own url to get responses back
    public init(
        host: URL = URL(string: "https://go.cb-w.com/wsegue")!,
        callback: URL
    ) {
        self.appId = Bundle.main.bundleIdentifier!
        
        self.host = host
        if callback.pathComponents.count < 2 { // [] or ["/"]
            self.callback = callback.appendingPathComponent("wsegue")
        } else {
            self.callback = callback
        }
        
        let sdkBundle = Bundle(for: Self.self)
        self.version = sdkBundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0"
    }
    
    // MARK: - Send message
    
    public func initiateHandshake(completion: @escaping (Result<Void,Error>) -> Void) {
        let message = RequestMessage(
            uuid: UUID(),
            sender: keyManager.ownPublicKey,
            content: .handshake(Handshake(appId: appId, callback: callback)),
            version: version
        )
        self.send(message) { result in
            switch result {
            case .success(let response):
                let result = self.handleNewSession(response)
                completion(result)
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    public func makeRequest(_ request: Request, onResponse: @escaping ResponseHandler) {
        let message = RequestMessage(
            uuid: UUID(),
            sender: keyManager.ownPublicKey,
            content: .request(request),
            version: version
        )
        self.send(message, onResponse)
    }
    
    private func send(_ message: RequestMessage, _ onResponse: @escaping ResponseHandler) {
        let url: URL
        do {
            url = try MessageConverter.encode(message, to: host, with: keyManager.symmetricKey)
        } catch {
            onResponse(.failure(error))
            return
        }
        
        UIApplication.shared.open(url, options: [.universalLinksOnly: true]) { result in
            guard result == true else {
                onResponse(.failure(CoinbaseWalletSDKError.openUrlFailed))
                return
            }
            
            self.taskManager.registerResponseHandler(for: message, onResponse)
        }
    }
    
    // MARK: - Receive message
    
    private func isWalletSegueMessage(_ url: URL) -> Bool {
        return url.host == callback.host && url.path == callback.path
    }
    
    public func handleResponse(_ url: URL) -> Result<Bool, Error> {
        guard isWalletSegueMessage(url) else {
            return .success(false)
        }
        
        let message: ResponseMessage
        do {
            message = try MessageConverter.decode(url, with: keyManager.symmetricKey)
        } catch {
            return .failure(error)
        }
        
        taskManager.handleMessage(message)
        
        return .success(true)
    }
    
    // MARK: - Session
    
    public func isConnected() -> Bool {
        return keyManager.symmetricKey != nil
    }
    
    public func resetConnection() -> Result<Void, Error> {
        do {
            try keyManager.resetOwnPrivateKey()
            return .success(())
        } catch {
            return .failure(error)
        }
    }
    
    private func handleNewSession(_ response: Response) -> Result<Void, Error> {
        guard
            let encodedPeerPublicKey = response.results.first,
            let data = Data(base64Encoded: encodedPeerPublicKey),
            let peerPublicKey = try? PublicKey(rawRepresentation: data)
        else {
            return .failure(CoinbaseWalletSDKError.decodingFailed)
        }
        
        do {
            try self.keyManager.storePeerPublicKey(peerPublicKey)
            return .success(())
        } catch {
            return .failure(error)
        }
    }
    
    // MARK: - static methods
    
    public static func isCoinbaseWalletInstalled() -> Bool {
        return UIApplication.shared.canOpenURL(URL(string: "cbwallet://")!)
    }
}

File: native/android/CoinbaseWalletNativeSDK/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt
package com.coinbase.android.nativesdk

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.coinbase.android.nativesdk.key.KeyManager
import com.coinbase.android.nativesdk.message.Content
import com.coinbase.android.nativesdk.message.Message
import com.coinbase.android.nativesdk.message.MessageConverter
import com.coinbase.android.nativesdk.message.Request
import com.coinbase.android.nativesdk.message.ResponseHandler
import com.coinbase.android.nativesdk.task.TaskManager
import java.security.interfaces.ECPublicKey
import java.util.UUID

const val CBW_APP_LINK = "go.cb-w.com"
const val CBW_PACKAGE_NAME = "org.toshi"
const val CBW_SCHEME = "cbwallet://"

typealias OpenIntentCallback = (Intent) -> Unit

class CoinbaseWalletSDK(
    domain: Uri,
    private val appContext: Context,
    private val hostPackageName: String = CBW_PACKAGE_NAME,
    private val openIntent: OpenIntentCallback
) {
    private val domain: Uri
    private val sdkVersion = BuildConfig.LIBRARY_VERSION_NAME
    private val keyManager by lazy { KeyManager(appContext, hostPackageName) }
    private val taskManager by lazy { TaskManager() }

    private val launchWalletIntent: Intent?
        get() = appContext.packageManager.getLaunchIntentForPackage(hostPackageName)

    val isWalletInstalled
        get() = launchWalletIntent != null

    init {
        this.domain = if (domain.pathSegments.size < 2) {
            domain.buildUpon()
                .appendPath("wsegue")
                .build()
        } else {
            domain
        }
    }

    fun initiateHandshake(initialRequest: Request.Request? = null, onResponse: ResponseHandler) {
        val message = Message(
            uuid = UUID.randomUUID().toString(),
            version = sdkVersion,
            sender = keyManager.ownPublicKey,
            content = Content.RequestContent(
                request = Request.Handshake(
                    domain = domain.toString(),
                    initialRequest = initialRequest
                )
            )
        )

        send(message, onResponse)
    }

    fun makeRequest(
        request: Request.Request,
        onResponse: ResponseHandler
    ) {
        val message = Message(
            uuid = UUID.randomUUID().toString(),
            sender = keyManager.ownPublicKey,
            version = sdkVersion,
            content = Content.RequestContent(request)
        )

        send(message, onResponse)
    }

    fun handleResponse(url: Uri): Boolean {
        if (!isWalletSegueResponseURL(url)) {
            return false
        }

        val ownPublicKey = keyManager.ownPublicKey
        val peerPublicKey = keyManager.peerPublicKey

        val message = MessageConverter.decode(
            url = url,
            ownPublicKey = ownPublicKey,
            ownPrivateKey = keyManager.ownPrivateKey,
            peerPublicKey = peerPublicKey
        )

        if (peerPublicKey == null && message.sender != ownPublicKey) {
            keyManager.storePeerPublicKey(message.sender as ECPublicKey)
        }

        return taskManager.handleResponse(message)
    }

    private fun send(message: Message, onResponse: ResponseHandler) {
        val uri = MessageConverter.encode(
            message = message,
            recipient = Uri.parse(CBW_SCHEME),
            ownPrivateKey = keyManager.ownPrivateKey,
            peerPublicKey = keyManager.peerPublicKey
        )

        val intent = launchWalletIntent
        if (intent == null) {
            onResponse(Result.failure(CoinbaseWalletSDKError.OpenWalletFailed))
            return
        }

        // Prevent intent from launching app in new window
        intent.type = Intent.ACTION_VIEW
        if (intent.flags and Intent.FLAG_ACTIVITY_NEW_TASK > 0) {
            intent.flags = intent.flags and Intent.FLAG_ACTIVITY_NEW_TASK.inv()
        }

        intent.data = uri

        taskManager.registerResponseHandler(message, onResponse)
        openIntent(intent)
    }

    private fun isWalletSegueResponseURL(uri: Uri): Boolean {
        return uri.host == domain.host && uri.path == domain.path && uri.getQueryParameter("p") != null
    }
}
